libraries/chain/include/steem/chain/database.hpp

```diff
@@ -364,6 +364,12 @@ namespace steem { namespace chain {
          /** this updates the vote of a single witness as a result of a vote being added or removed*/
          void adjust_witness_vote( const witness_object& obj, share_type delta );
 
+         /** this is called during block generation to ensure the witness votes for a hardfork that matches the binary configuration */
+         void adjust_witness_hardfork_version_vote( const witness_object& witness, signed_block& pending_block );
```

Comment:
Do these need to be done in database?

Reply:
I think `apply_pending_transactions()` should be in database because it manipulates `_pending_tx_session`.
On the other hand, `adjust_witness_hardfork_version_vote()` could probably be extracted out with some interfaces to access `_hardfork_versions` and `_hardfork_times`.

Reply:
These functions are now integrated in to the block_producer class.